source/android/uitestapp/src/androidTest/java/io.adaptivecards.uitestapp/TestHelpers.kt

```diff
@@ -13,39 +15,103 @@ import io.adaptivecards.uitestapp.ui.inputs.RetrievedInput
+            .perform(ViewActions.click())
     }
 
     internal fun goToInputsScreen() {
         Espresso.onView(ViewMatchers.withId(R.id.navigation_inputs)).perform(ViewActions.click())
     }
 
-    internal fun findInputInValidatedContainer(validatedContainerTagId : String) : ViewInteraction {
-        return Espresso.onView(ViewMatchers.withParent(ViewMatchers.withTagValue(Matchers.`is`(TagContent(validatedContainerTagId)))));
+    internal fun findInputInValidatedContainer(validatedContainerTagId: String): ViewInteraction {
+        return Espresso.onView(
+            ViewMatchers.withParent(
+                ViewMatchers.withTagValue(
+                    Matchers.`is`(
+                        TagContent(validatedContainerTagId)
+                    )
+                )
+            )
+        );
     }
 
-    internal fun clearTextInInput(input : ViewInteraction) {
+    internal fun clearTextInInput(input: ViewInteraction) {
         input.perform(ViewActions.scrollTo(), ViewActions.click(), ViewActions.clearText());
     }
 
-    internal fun setTextInInput(input : ViewInteraction, text : String) {
-        input.perform(ViewActions.scrollTo(), ViewActions.click(), ViewActions.clearText(), ViewActions.typeText(text));
+    internal fun setTextInInput(input: ViewInteraction, text: String) {
+        input.perform(
+            ViewActions.scrollTo(),
+            ViewActions.click(),
+            ViewActions.clearText(),
+            ViewActions.typeText(text)
+        );
+    }
+
+    internal fun selectPopupOption(optionText: String) {
+        val methodName = "SelectPopupOption";
+
+        val optionDataInteraction: DataInteraction =
+            Espresso.onData(Matchers.`is`(optionText)).inRoot(RootMatchers.isPlatformPopup());
+        Log.i(TAG, "$methodName - Found data $optionText");
+
+        optionDataInteraction.perform(ViewActions.scrollTo());
+        Log.i(TAG, "$methodName - Scrolled to element");
+
+        optionDataInteraction.check(ViewAssertions.matches(ViewMatchers.isDisplayed()));
+        Log.i(TAG, "$methodName - Asserted element is displayed");
+
+        optionDataInteraction.perform(ViewActions.click());
+        Log.i(TAG, "$methodName - Clicked on option $optionText");
     }
 
-    internal fun selectPopupOption(optionText : String) {
-        Espresso.onData(Matchers.`is`(optionText)).inRoot(RootMatchers.isPlatformPopup()).perform(ViewActions.scrollTo(), ViewActions.click());
+    internal fun pickItemInFilteredChoiceSet(
+        inputName: String,
+        hint: String,
+        choiceSetOption: String
+    ) {
+        val methodName = "PickItemInFilteredChoiceSet"
+        val maxRetries = 5
+        var retries = 0
+
+        while (retries < maxRetries) {
+            ++retries
+
+            Log.i(TAG, "$methodName - Try #$retries")
+
+            try {
+                setTextInInput(findInputInValidatedContainer(inputName), hint)
```

Comment:
log before action. or update the log info. For example: "Set value to $hint complete"
```                Log.i(TAG, "$methodName - Set value to $hint")
                setTextInInput(findInputInValidatedContainer(inputName), hint)

                Log.i(TAG, "$methodName - Set option to $choiceSetOption")
                selectPopupOption(choiceSetOption)
```